services/bulk_fix.py

```diff
@@ -0,0 +1,130 @@
+ 'Interview:SubjectAge:number': 'INTERVIEW:SUBJECT_AGE:number',
+ 'Interview:SubjectID': 'INTERVIEW:SUBJECT_ID',
+ 'Laxatives:Average_number_per_month': 'LAXATIVES:AVERAGE_NO_MONTH',
+ 'Laxatives:Average_number_per_week': 'LAXATIVES:AVERAGE_NO_WEEK',
+ 'OtherMethod:Average_number_per_month': 'OTHER_METHOD:AVERAGE_NO_MONTH',
+ 'OtherMethod:Average_number_per_week': 'OTHER_METHOD:AVERAGE_NO_WEEK',
+ 'OtherMethod:Name': 'OTHER_METHOD:NAME',
+ 'Vomiting:Average_number_per_month': 'VOMITING:AVERAGE_NO_MONTH',
+ 'Vomiting:Average_number_per_week': 'VOMITING:AVERAGE_NO_WEEK'}
+
+
+def find_all_questionnaires():
+###
+# find all distinct questionnaires by major_version
+# and return their major version and shortname
+###
+    client = pymongo.MongoClient('mongodb')
+    questionnaires = client.dsm.questionnaires
+    versions = questionnaires.aggregate([
+                {'$match': {'deleted_on': None}},
+                {'$sort': {'version.major': 1, 'version.minor': -1}},
+        {'$group': {
+                    '_id': '$version.major',
+                    'instrument_id': {'$first': '$instrument_id'},
+                    'name': {'$first': '$name'},
+                    'urlname': {'$first': '$urlname'},
+                    'minor': {'$first': '$version.minor'},
+                    'major': {'$first': '$version.major'},
+                    'shortname': {'$first': '$version.shortname'},
+                    'created_by': {'$first': '$created_by'},
+                    'created_on': {'$first': '$created_on'}
+                }},
+                {'$sort': {'created_on': 1}},
+                {'$group':{
+                    '_id': '$instrument_id',
+                    'instrument_id': {'$first': '$instrument_id'},
+                    'name': {'$first': '$name'},
+                    'urlname': {'$first': '$urlname'},
+                    'versions': {'$addToSet': {
+                        'version': {
+                            'created_by': '$created_by',
+                            'created_on': '$created_on',
+                            'shortname': '$shortname',
+                            'major': '$major',
+                            'minor': '$minor'
+                        }
+                    }},
+                }},
+                {'$sort': {'name': 1}}
+            ])
+    return versions.next()
+
+
+def do_sub(txt):
+    for m, sb in SUBS.items():
+        if txt and m in txt:
+            txt = re.sub(m, sb, txt)
+    return txt
+
+
+def make_url(version, debug=True):
+    params = {
+        'q': version['urlname'],
+        'major': version['major'],
+        minor: version['minor'],
+        debug: debug
+    }
+    ret = base64.b64encode(params)
+    return ret
+
+
+def parse_doc(questionnaire):
+    for q in questionnaire.get('questions', []):
+        q['probe_text'] = do_sub(q.get('probe_text', ""))
+    return questionnaire
+
+
+def bulk_permissions(user):
+    client = pymongo.MongoClient('mongodb')
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *client*. The resource is allocated by call *pymongo.MongoClient*. Execution paths that do not contain closure statements were detected. Close *client* in a try-finally block to prevent the resource leak.